src/Microsoft.TemplateEngine.Cli/TemplateResolution/ListOrHelpTemplateListResolutionResult.cs

```diff
@@ -0,0 +1,241 @@
+        {
+            get
+            {
+                if (_exactMatchedTemplates == null)
+                {
+                    _exactMatchedTemplates = GetCoreMatchedTemplatesWithDisposition(x => x.IsMatch);
+                }
+                return _exactMatchedTemplates;
+            }
+        }
+
+        /// <summary>
+        /// Returns list of exact or partially matched templates by name and exact match by language, filter, baseline (if specified in command paramaters) grouped by group identity
+        /// </summary>
+        public IReadOnlyCollection<IGrouping<string, ITemplateMatchInfo>> ExactMatchedTemplatesGrouped
+        {
+            get
+            {
+                return ExactMatchedTemplates.GroupBy(x => x.Info.GroupIdentity, x => !string.IsNullOrEmpty(x.Info.GroupIdentity)).ToList();
+            }
+        }
+
+        /// <summary>
+        /// Returns list of exact or partially matched templates by name and mismatch in any of the following: language, filter, baseline (if specified in command paramaters)
+        /// </summary>
+        public IReadOnlyCollection<ITemplateMatchInfo> PartiallyMatchedTemplates
+        {
+            get
+            {
+                if (_partiallyMatchedTemplates == null)
+                {
+                    _partiallyMatchedTemplates = GetCoreMatchedTemplatesWithDisposition(t => t.HasNameMatchOrPartialMatch() && t.HasAnyMismatch());
+                }
+                return _partiallyMatchedTemplates;
+            }
+        }
+
+        /// <summary>
+        ///  Returns list of exact or partially matched templates by name and mismatch in any of the following: language, filter, baseline (if specified in command paramaters); grouped by group identity
+        /// </summary>
+        public IReadOnlyCollection<IGrouping<string, ITemplateMatchInfo>> PartiallyMatchedTemplatesGrouped
+        {
+            get
+            {
+                return PartiallyMatchedTemplates.GroupBy(x => x.Info.GroupIdentity, x => !string.IsNullOrEmpty(x.Info.GroupIdentity)).ToList();
+            }
+        }
+
+
+        /// <summary>
+        /// Returns true when at least one template in unambiguous matches default language
+        /// </summary>
+        public bool HasUnambiguousTemplateGroupForDefaultLanguage
+        {
+            get
+            {
+                if (!HasUnambiguousTemplateGroup)
+                {
+                    return false;
+                }
+                return UnambiguousTemplatesForDefaultLanguage.Any();
+            }
+        }
+
+        /// <summary>
+        /// Returns collecion of templates from unamgibuous group that matches default language
+        /// </summary>
+        public IReadOnlyCollection<ITemplateMatchInfo> UnambiguousTemplatesForDefaultLanguage
+        {
+            get
+            {
+                return UnambiguousTemplateGroup.Where(t => t.HasDefaultLanguageMatch()).ToList();
```

Comment:
We are calling `ToList` in many places in this class. If we don't plan to enumerate each data multiple times, I think it would be better to just return an `IEnumerable<T>` in almost all of the cases for better memory and cpu efficiency.

Even if someone wants to re-enumerate the data multiple times, they can just call `ToList` themselves and pay the cost.

Reply:
I noticed it already however most of the methods that using that class expect collection to be returned, not IEnumerable. 
Also a lot of methods use `Count` and counting enumeration multiple times might cause performance loss. 

I can change that but that would result in touching almost each method in related classes, also due to many Count's not sure if we can get better performance. However I would agree that at least preserving list to private variable might be better then creating it each time.

Reply:
I would return to it after the roadmap meeting today, if we have an extra day I can go for refactoring all the classes to see where it makes sense to implement IEnumerable.